packages/sdk-v2/src/sdk/claim/utils/getRecentClaims.ts

```diff
@@ -15,80 +17,105 @@ export const getRecentClaims = async (
+        throw new Error("Failed to fetch events from explorer");
       }
+      return result.result;
     });
+    const fallbackResult = await fallback(apiCalls);
+    return fallbackResult;
+  };
 
-    if (callResults.length > 0) {
-      const transactionsSorted = orderBy(callResults, ["timeStamp"], ["desc"]);
-
-      const formatted = transactionsSorted.map((tx: any) => {
-        const { address, data, timeStamp, transactionHash } = tx;
-
-        const isGdPool = tx.topics[0] === claimHash;
-
-        const claimAmount = ethers.BigNumber.from(data);
-        const date = moment(ethers.BigNumber.from(timeStamp).toNumber() * 1000).utc();
-
-        // todo: add utilty getContractName(address) to read pool name from registry
-        // for now we assume its either GoodDollar or RedTent
-        const contractName = isGdPool ? "GoodDollar" : "RedTent";
+  const fetchLogsFromBlockchain = async (hash: string, pool: string) => {
+    const result = await library.getLogs({
+      address: pool,
+      topics: [hash, sender32],
+      fromBlock: "0x" + blockNoStart.toString(16),
+      toBlock: "0x" + lastBlockNo.toString(16)
+    });
 
-        return { address, claimAmount, contractName, date, transactionHash, isPool: !isGdPool };
-      });
+    if (!isArray(result)) throw new Error("Failed to fetch events from rpc");
 
-      return formatted;
-    }
+    const withTimestamp = await Promise.all(
+      result.map(async log => {
+        const b = await library.getBlock(log.blockNumber);
+        log.timeStamp = b.timestamp;
+        return log;
+      })
+    );
+    return withTimestamp;
+  };
+  try {
+    const results = await Promise.all(
+      flatten(
+        pools.map(pool =>
+          hashes.map(hash =>
+            fetchLogs(hash, pool).catch(() => {
+              return fetchLogsFromBlockchain(hash, pool);
+            })
+          )
+        )
+      )
+    );
+    const callResults: any[] = flatten(results).filter(_ => _);
+
+    const transactionsSorted = orderBy(callResults, ["blockNumber"], ["desc"]);
+
+    const formatted = transactionsSorted.map((tx: any) => {
+      const { address, data, timeStamp, transactionHash } = tx;
+
+      const isGdPool = tx.topics[0] === claimHash;
+
+      const claimAmount = ethers.BigNumber.from(data);
+      const date = moment(ethers.BigNumber.from(timeStamp).toNumber() * 1000).utc();
+
+      // todo: add utilty getContractName(address) to read pool name from registry
+      // for now we assume its either GoodDollar or RedTent
+      const contractName = isGdPool ? "GoodDollar" : "RedTent";
+
+      return { address, claimAmount, contractName, date, transactionHash, isPool: !isGdPool };
+    });
 
-    return callResults;
+    const merged = formatted.concat(claimsCache);
+    void AsyncStorage.setItem(cacheKey, { lastBlock: lastBlockNo, claimsCache: merged });
+    return merged;
   } catch (e) {
     throw new Error(`getRecentClaims -- Failed to fetch recent-claims: ${e}`);
```

Comment:
I think we can just return claims-cache (if length > 0) instead of always throwing error
(difference between infinite loader, or showing something)

or we need to add error handling in `ClaimContext`

Reply:
added returning of empty array in case of exception